### ISSUE TITLE
hide Artist, Genre and/or Year columns in Album Details View when the values are all the same

### ISSUE THREAD

**Issue by certuna** [navidrome/navidrome#846]
Currently, the WebUI desktop Album Details view looks like this:

![Screen Shot 2021-03-15 at 17 35 04](https://user-images.githubusercontent.com/62144283/111187944-d0493c00-85b4-11eb-893b-2a98f74a1d5e.png)

A couple of issues here:

1. the Artist for all tracks is the same, (edit: and the same as the Album Artist), so the Artist column is not needed
2. the Year however is different for each track (see the date range `1976-1989`), so it's pretty useful to have a Year column instead
3. the Navidrome UI doesn't show this, but the Genres are different for many tracks as well, so it's also pretty useful to have a Genre column _- to be discussed, is there enough space for that?_
 
In short: for these three fields, if the metadata is the same for all tracks, do not show the column, if it isn't, then show it.

Note: this only concerns the Desktop WebUI, for mobile (`isDesktop` = false), Year and Genre columns should probably be dropped.

This view is built by this JS: https://github.com/navidrome/navidrome/blob/847531391dab0dd4f5ac4951767d4698f920dfdb/ui/src/album/AlbumSongs.js . 

- [ ] Add an `isCompilation` check, if it is false, (edit: and track artist = album artist) then hide the Artist column
- [ ] Add an `AllSameYear` check, if true then hide Year column
- [ ] Add an `AllSameGenre` check, if true then hide Genre column

**Comment by shravani05** [comment]
@certuna I would like to work on this. Can you please assign me this issue?

**Comment by metalheim** [comment]
Your assumption for "1. the Artist for all tracks is the same, so the Artist column is not needed" is wrong. The Albumartist can differ completely from the track-individual Artist.
![image](https://user-images.githubusercontent.com/44159957/111445759-ffbe8c80-870b-11eb-9185-bb6a44c6d3cf.png)
Typically, you will see this when an artist features another artist for just one song.

This is even more true when you look at a "Compilation" Album. If you were to omit the "Artist" column, you wouldn't be able to know which artist a song is from.
![image](https://user-images.githubusercontent.com/44159957/111445934-2b417700-870c-11eb-99c1-7341d7314e5e.png)

**Comment by certuna** [comment]
> This is even more true when you look at a "Compilation" Album. If you were to omit the "Artist" column, you wouldn't be able to know which artist a song is from.

This is why I wrote _Add an isCompilation check, if it is, then display the Artist column_

But you are right for your Mos Def example, the check should be to only drop the column when "AllTrackArtistsTheSame = true" *and* "Track Artist = Album Artist".

**Comment by metalheim** [comment]
That's right, if it's an AllTrackArtistsTheSame Album, you don't need the column. 
But I'm not sure how much replacing it with another Value will impact user-experience. (Every other album could have different columns which could be confusing).

I'd rather prefer implementing SongSimpleList from #823 to make it visually more clear while not hiding any information. 🤝

**Comment by certuna** [comment]
> Every other album could have different columns which could be confusing
Most other servers/players do the same though - only show track artists when they're different to the album artists. I think that part is fairly intuitive.

You are right that sometimes showing year and genre columns and sometimes not may be a bit messy - but how would you propose to handle albums where tracks have different artists, different years and/or different genres? Not show this at all, like now? Always show all three columns Artist/Year/Genre in desktop?

**Comment by armedev** [comment]
> > Every other album could have different columns which could be confusing
> 
> A lot of other servers/players (Plex for example, most mobile players like the Music app on iOS) do the same though - only show track artists for compilations (ie, when they're different to the album artist). I think that part is fairly intuitive, and also desirable especially for a mobile SongSimpleList.
> 
> You are right that sometimes showing year and genre columns and sometimes not may be a bit messy - but how would you propose to handle albums where tracks have different artists, different years and/or different genres? Not show this at all, like now? Always show all three columns Artist/Year/Genre in desktop?
> 
> The SongSimpleList from #823 has nothing to do with this in principle, that is for the mobile layout, this is for the desktop layout:
> 
> > for mobile (isDesktop = false), Year and Genre columns should probably be dropped.

yeah, @certuna is right SongSimpleList component is for simple and smaller mobile displays, and this has been created to deal with a new design for song components in desktop versions.

**Comment by armedev** [comment]
@certuna if we completely remove the artist column and if we have any additional artist remaining we can append it to the end of the title with `ft.` as a prefix.

like if you see here

![Screenshot from 2021-03-17 12-24-42](https://user-images.githubusercontent.com/44159957/111446402-a6a32880-870c-11eb-87e4-2cb94d7c5c37.png)

in the first song, it is already prefixed we just check for "ft." in the title if not we append `ft. artistName`
if 'ft.' is present we just display it


how does this look like @deluan @jvoisin?

**Comment by certuna** [comment]
This is not so trivial.

First of all, Navidrome currently doesn't support multiple values in the artist field yet. Nor in the genre field, for that matter. What you see in that example is stored in Navidrome as one artist "Mos Def/Cassidy", not two artists "Mos Def" and "Cassidy". Aside from client-side parsing/splitting of metadata fields (a minefield in itself, it's already hard enough to do that server-side) there's no easy way to strip "/Cassidy" and move it to the song title.

Secondly, I'm not so sure if moving the second (and third/etc?) artists to an appended string `Song Title (ft. SecondArtist & ThirdArtist)` in the song title field is desirable at all. In most cases, the song title is already longer than the artist name already `I Would Do Anything For Love (But I Won't Do That) (DJ Awesome Remix)`, and a big part of the advantage of supporting multiple artists is that you can show them listed in the artist field - so why take them out again?

I would propose, if there's even one song with a second artist, we show the Artist column. Only when *all* track artists are identical to the album artist, then hide the Artist column.

**Comment by shravani05** [comment]
> > This is even more true when you look at a "Compilation" Album. If you were to omit the "Artist" column, you wouldn't be able to know which artist a song is from.
> 
> This is why I wrote in the opening post _Add an isCompilation check, if it is, then display the Artist column_
> 
> But you are right for your Mos Def example, the check should be to only drop the column when "AllTrackArtistsTheSame = true" _and_ "Track Artist = Album Artist".
> 
> @shravani05 I don't think I can assign issues - can you contact @jvoisin or @deluan on this? Also, probably good to have a bit more discussion/consensus on the desired behaviour than just two days before we start coding :)

@certuna no problem :)

**Comment by aldrinjenson** [comment]
Hey, I've a suggestion: Why don't we let the user decide? As in how nice it would it be if the user has the option to display the fields/attributes of his choice. What I have in mind is a button near the filter and shuffle options which opens a modal  on click from which the user could select the fields(check-boxes) to be displayed in each song row. 
I think this could also help if in the future the user wants to also see fields like genre, file size etc as well (which currently is possible only on clicking each song separately in the drop down)
![image](https://user-images.githubusercontent.com/53407417/112721071-f4e3d300-8f27-11eb-8dd7-f19d9757d7b7.png)

P.S. This feature is pretty common in most desktop music clients I've used
eg: Here's how a similar feature is implemented in Rhythmbox
![image](https://user-images.githubusercontent.com/53407417/112721214-be5a8800-8f28-11eb-9611-42975b688799.png)

I could try to make a pr demo on this if this suggestion is worth considering.

**Comment by deluan** [comment]
Yes, this is something I want to implement at some point. [React-Admin Enterprise Edition](https://marmelab.com/ra-enterprise/modules/ra-preferences#selectcolumnsbutton-store-datagrid-columns-in-preferences) has this nicely implemented, which design we could "borrow": 

https://user-images.githubusercontent.com/331353/112741220-a3e8d480-8f51-11eb-8a3a-799f97cc375a.mov

**Comment by aldrinjenson** [comment]
Hey @deluan and @certuna , I've made a pr related to the change I suggested. Ig the original issue was different, tho please do take a look and see if this [pr](https://github.com/navidrome/navidrome/pull/923) is helpful in that direction. 🤞

**Comment by certuna** [comment]
I think this configurable columns thing works in the List view of Songs, Albums, or Artists, but that's something for another ieeus. The idea of this issue is to dynamically hide redundant info from the album details. You don't want the user to have to change the columns setting every time to show the Year column when it's a compilation, and then manually hide it again when it's a regular album - that behaviour should be automatic.

**Comment by aldrinjenson** [comment]
Got it. My bad. Got muddled up with the issues it seems.😬 So I think, it'd then be better then to create a new issue separately for the menus.

**Comment by deluan** [comment]
@aldrinjenson No need to create a separated issue. I'll review your PR, just not close this issue

**Comment by certuna** [comment]
Still relevant and valuable, on my To Do list
